completions/_abbr

```diff
@@ -10,197 +10,196 @@
+          _arguments \
+            '(-S --session -U --user)'{-S,--session}'[available in this session]' \
+            '(-U --user -S --session)'{-U,--user}'[available in all sessions]' \
+            '(-r --regular -g --global)'{-r,--regular}'[expand at the start of the line]' \
+            '(-g --global -r --regular)'{-g,--global}'[expand everywhere]' \
+            '(--dry-run)--dry-run[see what would result, without making any actual changes]' \
+            '(-q --quiet -qq --quieter)'{-q,--quiet}'[silence success output]' \
+            '(-qq --quieter -q --quiet)'{-qq,--quieter}'[silence success output and warnings]' \
+            '(-f --force)'{-f,--force}'[ignore warnings]' \
+            '(--file)--file[path to a Git config file]:filename:_files -/' \
+            '(--prefix)--prefix[prefix added to the ABBREVIATIONs]'
+          ret=0
+          ;;
+        R|\
+        rename)
+          # [<SCOPE>] [<TYPE>] [--dry-run] [(--quiet | --quieter)] [(-f | --force)] OLD NEW
+          _arguments \
+            '(-S --session -U --user)'{-S,--session}'[available in this session]' \
+            '(-U --user -S --session)'{-U,--user}'[available in all sessions]' \
+            '(-r --regular -g --global)'{-r,--regular}'[expand at the start of the line]' \
+            '(-g --global -r --regular)'{-g,--global}'[expand everywhere]' \
+            '(--dry-run)--dry-run[see what would result, without making any actual changes]' \
+            '(-q --quiet -qq --quieter)'{-q,--quiet}'[silence success output]' \
+            '(-qq --quieter -q --quiet)'{-qq,--quieter}'[silence success output and warnings]' \
+            '(-f --force)'{-f,--force}'[ignore warnings]' \
+            '1: :__abbr_abbreviations'
+          ret=0
+          ;;
+      esac
+      ;;
+  esac
+
+  return ret
+}
+
+(( $+functions[__abbr_extract_abbreviations] )) ||
+__abbr_extract_abbreviations() {
+  local -a tmp_abbr tmp_expd
+  local scope type
+
+  if (( $words[(Ie)-U] || $words[(Ie)--user] )); then
+    scope='-U'
+  elif (( $words[(Ie)-S] || $words[(I)--session] )); then
+    scope='-S'
+  elif (( $words[(Ie)-r] || $words[(Ie)--regular] )); then
+    type='-r'
+  elif (( $words[(Ie)-g] || $words[(Ie)--global] )); then
+    type='-g'
+  fi
+
+  tmp_abbr=(${${(f)"$(_call_program abbreviations abbr list-abbreviations $scope $type 2>/dev/null)"}//\"/})
+  tmp_expd=(${${${${(f)"$(_call_program abbreviations abbr list $scope $type 2>/dev/null)"}#*=}#\"}%\"})
+
+  if (( ${#ABBR_REGULAR_USER_ABBREVIATIONS} )); then
+    local len=${#tmp_abbr}
+    for (( i = 1; i <= len; i++ )); do
+      abbreviations+=(${tmp_abbr[$i]}:${tmp_expd[$i]})
+    done
+  else
+    abbreviations=()
+  fi
+}
+
+(( $+functions[__abbr_abbreviations] )) ||
+__abbr_abbreviations() {
+  local -a abbreviations
+  __abbr_extract_abbreviations
+
+  _describe -t abbreviations 'abbr abbreviations' abbreviations
+}
 
-return ret
+_abbr
```

Comment:
> Is there precedent for wrapping this in a function? (it's been a while since I studied others' completion files) Is it advantageous?

There's a lot of discrepancies for the structure of completions as far as I can see. So my reasoning was that `return` is valid inside functions.

I've read mainly these completions:

- [`/usr/share/zsh/functions/Completion/Unix/_git`](https://github.com/zsh-users/zsh/blob/master/Completion/Unix/Command/_git)
- [`/usr/share/zsh/functions/Completion/Unix/_pkgadd`](https://github.com/zsh-users/zsh/blob/master/Completion/Unix/Command/_pkgadd)
- [`/usr/share/zsh/functions/Completion/Unix/_rar`](https://github.com/zsh-users/zsh/blob/master/Completion/Unix/Command/_rar)
- [`/usr/share/zsh/vendor-completions/_docker`](https://github.com/docker/cli/blob/master/contrib/completion/zsh/_docker)

In some of those files they don't even return anything or wrap the logic in a main function `_<command_name>()`, some of them do. It seems that when there's some complex logic behind the completions they tend to use a wrapper function. 

---

> Any risk of confusing the shell by giving it the same name as the file? at one point I had a function in zsh-abbr.zsh called _abbr, and that tripped up the completion system.

None as far as I know if it's only used in the completion file, this is used by git and docker completions (from the examples above)

Reply:
I unwrapped the main logic and place the helper functions above it. It seems to work fine. If you want we can do this.

Reply:
Undecided.

I feel like zsh completions files are already so opaque, simplify as much as possible so give the dev community examples that aren't impossible to understand. And it's been a while since I looked, but I'm pretty sure I've seen a bunch of others that don't wrap. On the other hand, my convention is to wrap shell script file contents in a function!

Let's come back to it after everything else is finalized.

Reply:
I think we can click resolve on this. If you agree @Farid-NL, go ahead and do it